guardian/mobile-n10n
# Collect token delivery latency metrics

cc @michaelwmcnamara (who also worked on this PR)

## What does this change?

This PR allows us to calculate (and collect CloudWatch metrics for) token delivery latency.

This means that we can use a standard latency metric in CloudWatch to visualise internal stack performance (i.e. we can check how the stack is performing in isolation of Firebase/APNS delivery times, which may be unpredictable). For example, we can calculate the 99th, 95th & 90th percentiles for token delivery latency. 

![image](https://user-images.githubusercontent.com/19384074/208446675-d5163adb-1edd-4b90-9832-34588babbb13.png)

We can also calculate an SLI based on this metric, for example:

Good events = _the number of token deliveries where the delivery took less than X seconds_
Valid events = _the number of token deliveries_
SLI = _(the number of token deliveries where the delivery took less than X seconds / the number of token deliveries) * 100_ 

IIUC a single token delivery directly corresponds to an individual user receiving a notification[^1]. Consequently, using token delivery latency is likely to be useful for an SLI (and potentially an SLO) because:

1. It is easy to understand/reason about the correlation between latency increasing and user happiness decreasing[^2]
2. It allows us to easily compare a (new) token delivery SLI with the (existing) 90in2 SLI, which also operates at the level of individual-user deliveries

## How to test

* I have deployed this to `CODE` and confirmed that the latency metrics being calculated look sensible
* Since this PR changes the JSON that is passed between the various microservices, I have also confirmed that:
  * A (dry-run) notification can be sent if `notification` is running an old build and the `workers` are running this new build
  * A (dry-run) notification can be sent if `notification` is running this new build and the `workers` are still running an old build
* I sent a single 'standard' notification (i.e. not a dry-run) to confirm that metrics were actually being pushed to CloudWatch. There were not many datapoints for this; I assume that we must have thousands of 'fake' tokens in `CODE`?
* We have added unit tests for the latency metric batching logic, as this was a little tricky to write

## How can we measure success?

* We can better visualise performance of the stack via CloudWatch (and Grafana)
* We should also be able to create alerts related to this metric/SLI (if desired)

## Have we considered potential risks?

I can think of a few potential risks here:

1. We are adding a new field to the JSON that is passed between the different microservices (see https://github.com/guardian/mobile-n10n/commit/a8d26d171c249e229a2240bf472d9d183866d3c2). I have tried to mitigate this risk making the field optional (for consumers) in the first instance. This means that any messages that are already in the queue(s) should still be processed correctly, even though they lack this new field. I have also tested (see above) to confirm that the service can continue operating successfully even if this build is rolled out to different microservices at different times. Once this PR has been deployed, we'll be able to make the new field mandatory in all of the models and this complexity will be removed.

1. We are making more API calls to CloudWatch as a result of this PR. This has the potential to:
   1. Slow the worker function(s) down 
   1. ~Introduce invocation failures (e.g. due to a CloudWatch API call failing).~ Update: I don't think a CloudWatch call failing will cause the invocation to fail, see [this comment](https://github.com/guardian/mobile-n10n/pull/884#discussion_r1052310294).
   1. Increase AWS costs[^3]
  
In order to mitigate the above, we have implemented batching logic so that latency metrics are pushed to CloudWatch in batches of _150 unique values_. 

I think each worker is currently [processing 1000 tokens](https://github.com/guardian/mobile-n10n/blob/471357ac709baada6dc9cfb6830a4bdc88c3e4b3/notificationworkerlambda/src/main/scala/com/gu/notifications/worker/Harvester.scala#L67)? So in theory a worker invocation could result in 7 additional API calls.

However, as a unique value in this context is the number of seconds spent processing a notification and Lambda invocations are well below 150 seconds, I would expect each Lambda to make one additional API call in practice[^4].

Note that there are options for optimising further here (e.g. we could move to [embedded metric format](https://aws.amazon.com/about-aws/whats-new/2019/11/amazon-cloudwatch-launches-embedded-metric-format/), or combine [some of the other CloudWatch monitoring functionality](https://github.com/guardian/mobile-n10n/blob/471357ac709baada6dc9cfb6830a4bdc88c3e4b3/notificationworkerlambda/src/main/scala/com/gu/notifications/worker/utils/Cloudwatch.scala#L45-L77) into the same API calls).

[^1]: Assuming that the Firebase/APNS delivery succeeds eventually.
[^2]: I think it will be easier to explain this SLI to stakeholders than something more technical (e.g. throughput).
[^3]: Since we [pay per API call](https://aws.amazon.com/cloudwatch/pricing/).
[^4]: For example, when testing in `CODE` I observed that the Android Worker makes a single API call where there is only unique value - the number of seconds spent processing  (e.g. 45 seconds) - and the count is the number of tokens processed in that time (e.g. 1000).
